src/frequenz/sdk/config/_util.py

```diff
@@ -4,22 +4,30 @@
 """Utilities to deal with configuration."""
 
 from collections.abc import Mapping
-from typing import Any, TypeVar, cast
+from typing import Any, ClassVar, Protocol, TypeVar, cast
 
 from marshmallow import Schema
 from marshmallow_dataclass import class_schema
 
-T = TypeVar("T")
+
+class Dataclass(Protocol):
+    """A protocol for dataclasses."""
+
+    __dataclass_fields__: ClassVar[dict[str, Any]]
+    """The fields of the dataclass."""
```

Comment:
It took me a while to understand why it checks if ` __dataclass_fields__` is available .
Is it because is_dataclass method uses it?
Are you sure we can rely on this field?

Reply:
> It took me a while to understand why it checks if  __dataclass_fields__ is available .

Good point, I will add a comment.

>  Is it because is_dataclass method uses it?

No, I added the protocol to be able to do that, but the protocol is a way to tell the type system you want to accept dataclasses. Not sure how familiar are you with protocols but they are basically interfaces, [structural subtyping](https://mypy.readthedocs.io/en/stable/protocols.html) is called.

As dataclasses don't inherit from any class, we can't use a base class as the type for arguments for functions. `is_dataclass` is a dynamic check. This is sort of a hack, and assumes that anything that defines __dataclass_fields__ is a dataclass, which is not strictly true, but I think it should work pretty well in practice.

>  Are you sure we can rely on this field?

Well, if you are going to go legal, then no, we can't rely on it, at least according to [this](https://github.com/python/mypy/issues/15974#issuecomment-1694781006). I think in practice [everybody seems to be relying on it](https://github.com/python/mypy/issues/15974#issuecomment-1694993493) so I think we should be safe to do so for the time being.

Reply:
Added a comment to try to clarify this a bit.

Reply:
> Maybe instead we could check that in load_config method during runtime? 

I just checked and dataclass.is_dataclass returns True if class is  marshmallow_dataclass.dataclass...